[CSS Math Functions] De-duplicate `CSSCalcOperationNode::evaluateOperator` & `CalcExpressionOperation::evaluate`

nt1m opened the pull request:
#### 4ea2d50a5f28aa5b54ae68064d9421f3cc0b1060
<pre>
[CSS Math Functions] De-duplicate `CSSCalcOperationNode::evaluateOperator` &amp; `CalcExpressionOperation::evaluate`
<a href="https://bugs.webkit.org/show_bug.cgi?id=259021">https://bugs.webkit.org/show_bug.cgi?id=259021</a>
rdar://111962034

Reviewed by Darin Adler.

Use a common `evaluateCalcExpression` template function.

* Source/WebCore/css/calc/CSSCalcOperationNode.cpp:
(WebCore::CSSCalcOperationNode::evaluateOperator):
(WebCore::getNearestMultiples): Deleted.
* Source/WebCore/platform/calc/CalcExpressionOperation.cpp:
(WebCore::CalcExpressionOperation::evaluate const):
(WebCore::getNearestMultiples): Deleted.
* Source/WebCore/platform/calc/CalcOperator.h:
(WebCore::evaluateCalcExpression):

Canonical link: <a href="https://commits.webkit.org/265888@main">https://commits.webkit.org/265888@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/f6a5974f42a202e6248496413d442bc1e77cedc1

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/12162 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/12508 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/12810 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/13908 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/11736 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/12193 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/14924 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/12523 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/14419 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/12327 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/15/builds/13132 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/10300 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/14324 "Built successfully") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/39/builds/10418 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/10/builds/11041 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/18155 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/11499 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/36/builds/11201 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/14376 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/11692 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/9641 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/10908 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/15234 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/1359 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/11545 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

darinadler commented on Source/WebCore/css/calc/CSSCalcOperationNode.cpp:
```diff
@@ -1273,229 +1272,11 @@ bool CSSCalcOperationNode::equals(const CSSCalcExpressionNode& exp) const
-    }
-    ASSERT_NOT_REACHED();
-    return 0;
+    return evaluateCalcExpression<double>(calcOperator, children, [](double child) {
```
Why can’t the compiler deduce the template argument from the type of `children`?

nt1m replied:
Deduction works here!

darinadler commented on Source/WebCore/platform/calc/CalcOperator.h:
```diff
@@ -63,4 +65,239 @@ enum class CalcOperator : uint8_t {
 TextStream& operator<<(TextStream&, CalcOperator);
 
+template <typename T, typename Function>
+double evaluateCalcExpression(CalcOperator calcOperator, const Vector<T>& children, Function&& evaluate)
```
If we are not using deduction I suggest `Span<const T>` or `Span<T>`.

darinadler commented on Source/WebCore/platform/calc/CalcOperator.h:
```diff
@@ -63,4 +65,239 @@ enum class CalcOperator : uint8_t {
+        return sum;
+    }
+    case CalcOperator::Subtract:
+        ASSERT(children.size() == 2);
```
Not new: I don’t understand why some of these check sizes but others assert. Would be safer to always check and return NaN. Or more efficient to check elsewhere and assert here. But both depending on the operator?

nt1m replied:
I don't have much history on why the code is written this way, Nikos & Simon wrote most of this code, so I can ask them next week.

darinadler commented on Source/WebCore/platform/calc/CalcOperator.h:
```diff
@@ -63,4 +65,239 @@ enum class CalcOperator : uint8_t {
+        double sum = 0;
+        for (auto& child : children) {
+            auto evaluatedChild = evaluate(child);
+            if (std::isnan(evaluatedChild))
```
Not new: We can probably remove this without any effect on the results because of how NaN works. Would slightly speed up non-NaN cases and slow down NaN cases. Not important to change.

nt1m replied:
Removed.

darinadler commented on Source/WebCore/platform/calc/CalcOperator.h:
```diff
@@ -63,4 +65,239 @@ enum class CalcOperator : uint8_t {
+            return std::signbit(valueToRound) ? -0.0 : std::numeric_limits<double>::infinity();
+        }
+        auto ret = getNearestMultiples(valueToRound, roundingInterval);
+        return ret.second;
```
A local named `ret` is not WebKit coding style. In this case I would suggest merging into a single line. In other cases a name might be better.

nt1m replied:
Fixed

darinadler commented on Source/WebCore/platform/calc/CalcOperator.h:
```diff
@@ -63,4 +65,239 @@ enum class CalcOperator : uint8_t {
+        if (!std::isinf(valueToRound) && std::isinf(roundingInterval))
+            return std::signbit(valueToRound) ? -0.0 : +0.0;
+        auto ret = getNearestMultiples(valueToRound, roundingInterval);
+        auto upperB = ret.second;
```
`auto [lower, upper] = nearest` would be better.

nt1m replied:
Fixed

darinadler reviewed (approved):
A few ideas for tweaks. None mandatory.

nt1m committed:
[CSS Math Functions] De-duplicate `CSSCalcOperationNode::evaluateOperator` & `CalcExpressionOperation::evaluate`
https://bugs.webkit.org/show_bug.cgi?id=259021
rdar://111962034

Reviewed by Darin Adler.

Use a common `evaluateCalcExpression` template function.

* Source/WebCore/css/calc/CSSCalcOperationNode.cpp:
(WebCore::CSSCalcOperationNode::evaluateOperator):
(WebCore::getNearestMultiples): Deleted.
* Source/WebCore/platform/calc/CalcExpressionOperation.cpp:
(WebCore::CalcExpressionOperation::evaluate const):
(WebCore::getNearestMultiples): Deleted.
* Source/WebCore/platform/calc/CalcOperator.h:
(WebCore::evaluateCalcExpression):

Canonical link: https://commits.webkit.org/265888@main

darinadler commented on Source/WebCore/platform/calc/CalcOperator.h:
```diff
@@ -25,7 +25,9 @@
 
 #pragma once
 
+#include <cmath>
```
One post-landing thought. Is this header included in other headers? If so we might not want to put a large function template into it.

nt1m replied:
It is (although they're pretty fast to compile): https://searchfox.org/wubkat/search?q=CalcOperator.h&path=&case=false&regexp=false

It felt like a natural fit for the function template since it uses CalcOperator, but I'm open to other suggestions. A new file would work too, although I'm not sure what the name of that file would be.
